src/wayland/dmabuf/mod.rs

```diff
@@ -343,38 +250,60 @@ impl<H: DmabufHandler> ParamsHandler<H> {
             trace!(self.log, "Killing client providing bogus dmabuf buffer params.");
             return;
         }
-        let info = BufferInfo {
-            planes: ::std::mem::replace(&mut self.pending_planes, Vec::new()),
-            width,
-            height,
+
+        let mut buf = Dmabuf::new(
+            width as u32,
+            height as u32,
             format,
-            flags: BufferFlags::from_bits_truncate(flags),
+            DmabufFlags::from_bits_truncate(flags),
+        );
+        let planes = ::std::mem::replace(&mut self.pending_planes, Vec::new());
+        for (i, plane) in planes.into_iter().enumerate() {
+            let offset = plane.offset;
+            let stride = plane.stride;
+            let modi = plane.modifier;
+            buf.add_plane(plane.into_raw_fd(), i as u32, offset, stride, modi);
+        }
+        let dmabuf = match buf.build() {
+            Some(buf) => buf,
+            None => {
+                params.as_ref().post_error(
+                    ParamError::Incomplete as u32,
+                    format!("Provided buffer is incomplete, it has zero planes"),
+                );
+                return;
+            }
         };
-        let mut handler = self.handler.borrow_mut();
-        if let Ok(data) = handler.validate_dmabuf(info) {
-            if let Some(buffer) = params
-                .as_ref()
-                .client()
-                .and_then(|c| c.create_resource::<wl_buffer::WlBuffer>(1))
-            {
-                let buffer = handler.create_buffer(data, buffer);
-                trace!(self.log, "Creating a new validated dma wl_buffer.");
+
+        if let Some(buffer) = params
+            .as_ref()
+            .client()
+            .and_then(|c| c.create_resource::<wl_buffer::WlBuffer>(1))
+        {
+            buffer.as_ref().user_data().set_threadsafe(|| dmabuf);
+            buffer.quick_assign(|_, _, _| {});
+
+            let mut handler = self.handler.borrow_mut();
+            if handler(&buffer, ddata) {
+                trace!(self.log, "Created a new validated dma wl_buffer.");
                 params.created(&buffer);
+            } else {
+                trace!(self.log, "Refusing creation of an invalid dma wl_buffer.");
+                params.failed();
+                buffer.release();
```

Comment:
This is incorrect, we cannot send an event to the client associated with an object that it has never seen. In this case, the `WlBuffer` should only be created if the validation passes. Otherwise we'd be messing up the Wayland protocol state.

Reply:
Alright, I have rebased this on top of #274 (yes they are becoming increasingly dependent....), because with the changes from #288 I was able to solve this problem more easily. The new `ImportDma` trait now can also import a `Dmabuf`, that is not yet used by a `WlBuffer`. This allows the necessary import check to be performed without creating a `WlBuffer` first.